### PR TITLE
Add SCP timeout metrics

### LINKED ISSUE #1877: metrics: add metrics that track SCP timeouts
As per #1831, we're missing metrics that track how many times timeouts occur per SCP round for the various types of timers that exist (right now ballot protocol and nomination timers)
Easiest might be to have one metric per timer type, ie `scp.timeouts-X` where X is "ballot" or "nomination".
Alternative, that might be more useful is to capture those counters only at specific times:
* nomination timeouts before first prepare
* ballot timeouts before externalization

### PR DESCRIPTION
**Pull request by marta-lokhova**
Resolves #1877

### CODE REVIEW

File: src/herder/HerderSCPDriver.cpp [hunk 1]
```diff
@@ -469,6 +474,30 @@ HerderSCPDriver::timerCallbackWrapper(uint64_t slotIndex, int timerID,
...
+        if (SCPTimingIt != mSCPExecutionTimes.end())
+        {
+            auto& SCPTiming = SCPTimingIt->second;
+            if (timerID == Slot::BALLOT_PROTOCOL_TIMER)
+            {
+                if (SCPTiming.mSCPProgressState ==
+                    SCPProgressState::FIRST_PREPARE)
+                {
+                    // Timeout happened in between first prepare and externalize
+                    mSCPMetrics.mPrepareTimeout.Mark();
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: I think the metrics as measured are not going to be super useful: we should keep track of them in units of ledger (so maybe record the number of timeouts when we externalize using a histogram, the same way we track number of transactions/ledger)

File: src/herder/HerderSCPDriver.cpp [hunk 2]
```diff
@@ -897,6 +928,7 @@ HerderSCPDriver::recordSCPExecutionMetrics(uint64_t slotIndex)
     }
 
     auto& SCPTiming = SCPTimingIt->second;
+    SCPTiming.mSCPProgressState = SCPProgressState::EXTERNALIZE;
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: by moving to "per ledger", you won't need to keep track of EXTERNALIZE anymore as this function will be the only place where you interact with metrics

File: src/herder/HerderSCPDriver.cpp [hunk 3]
```diff
@@ -469,6 +474,30 @@ HerderSCPDriver::timerCallbackWrapper(uint64_t slotIndex, int timerID,
...
+                if (SCPTiming.mSCPProgressState ==
+                    SCPProgressState::FIRST_PREPARE)
+                {
+                    // Timeout happened in between first prepare and externalize
+                    mSCPMetrics.mPrepareTimeout.Mark();
+                }
+            }
+            else
+            {
+                if (SCPTiming.mSCPProgressState ==
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: you should just use `mPrepareStart` as indicator that prepare happened

**Comment by marta-lokhova** [comment]
updated

**Comment by MonsieurNicolas** [comment]
r+ bfa9e59bd1875ed482e79779e9eac6fd90f6a6fc

**Pull request merged**